File: frontend/src/utils/fileValidation.js
// File validation constants
export const MAX_FILE_SIZE = 10 * 1024 * 1024; // 10MB
export const MAX_IMAGE_SIZE = 5 * 1024 * 1024; // 5MB

export const ALLOWED_FILE_TYPES = {
  "application/pdf": [".pdf"],
  "application/msword": [".doc"],
  "application/vnd.openxmlformats-officedocument.wordprocessingml.document": [".docx"],
  "text/plain": [".txt"],
  "image/jpeg": [".jpg", ".jpeg"],
  "image/png": [".png"],
};

export const ALLOWED_IMAGE_TYPES = {
  "image/jpeg": [".jpg", ".jpeg"],
  "image/png": [".png"],
  "image/gif": [".gif"],
  "image/webp": [".webp"],
};

/**
 * Generic file validation helper
 * @param {File} file - The file to validate
 * @param {number} maxSize - Maximum file size in bytes
 * @param {Object} allowedTypes - Object mapping MIME types to extensions
 * @param {string} typeDescription - Human-readable description of allowed types
 * @returns {string|null} Error message if invalid, null if valid
 */
export const validateFileGeneric = (file, maxSize, allowedTypes, typeDescription) => {
  if (!file) {
    return "No file selected";
  }

  // Check file size
  if (file.size > maxSize) {
    return `File size must be less than ${maxSize / (1024 * 1024)}MB (selected file is ${(file.size / (1024 * 1024)).toFixed(2)}MB)`;
  }

  // Check file type by MIME type first
  let isValidType = Object.keys(allowedTypes).includes(file.type);

  // If MIME type check fails or is empty, check by file extension
  if (!isValidType || !file.type) {
    const fileName = file.name.toLowerCase();
    const dotIndex = fileName.lastIndexOf(".");
    if (dotIndex === -1) {
      return `Invalid file type. Allowed types: ${typeDescription}`;
    }
    const fileExtension = fileName.substring(dotIndex);
    const allowedExtensions = Object.values(allowedTypes).flat();
    isValidType = allowedExtensions.includes(fileExtension);
  }

  if (!isValidType) {
    return `Invalid file type. Allowed types: ${typeDescription}`;
  }

  return null; // Valid file
};

/**
 * Validate recipe document files (PDF, DOC, DOCX, TXT, JPG, PNG)
 * @param {File} file - The file to validate
 * @returns {string|null} Error message if invalid, null if valid
 */
export const validateRecipeDocument = (file) => {
  return validateFileGeneric(
    file,
    MAX_FILE_SIZE,
    ALLOWED_FILE_TYPES,
    "PDF, DOC, DOCX, TXT, JPG, PNG"
  );
};

/**
 * Validate image files (JPG, PNG, GIF, WEBP)
 * @param {File} file - The file to validate
 * @returns {string|null} Error message if invalid, null if valid
 */
export const validateImage = (file) => {
  return validateFileGeneric(
    file,
    MAX_IMAGE_SIZE,
    ALLOWED_IMAGE_TYPES,
    "JPG, PNG, GIF, WEBP"
  );
};

/**
 * Validate recipe image files (JPG, PNG, GIF, WEBP)
 * @param {File} file - The file to validate
 * @returns {string|null} Error message if invalid, null if valid
 * @deprecated Use validateImage instead
 */
export const validateRecipeImage = (file) => {
  return validateImage(file);
};

/**
 * Validate collection image files (JPG, PNG, GIF, WEBP)
 * @param {File} file - The file to validate
 * @returns {string|null} Error message if invalid, null if valid
 * @deprecated Use validateImage instead
 */
export const validateCollectionImage = (file) => {
  return validateImage(file);
};

/**
 * Format file size for display
 * @param {number} bytes - File size in bytes
 * @returns {string} Formatted size (e.g., "2.5MB")
 */
export const formatFileSize = (bytes) => {
  if (bytes === 0) return "0 Bytes";
  const k = 1024;
  const sizes = ["Bytes", "KB", "MB", "GB"];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return Math.round(bytes / Math.pow(k, i) * 100) / 100 + sizes[i];
};

File: backend/Utilities/FileValidation.cs
namespace RecipeManager.Api.Utilities;

/// <summary>
/// Centralized file validation utility for consistent file type validation across the application.
/// </summary>
public static class FileValidation
{
    /// <summary>
    /// Allowed content types for recipe document uploads (PDF, DOC, DOCX, TXT, JPG, PNG).
    /// </summary>
    public static readonly HashSet<string> AllowedDocumentTypes = new(StringComparer.OrdinalIgnoreCase)
    {
        "application/pdf",
        "application/msword",
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "text/plain",
        "image/jpeg",
        "image/png"
    };

    /// <summary>
    /// Allowed file extensions for recipe document uploads.
    /// </summary>
    public static readonly HashSet<string> AllowedDocumentExtensions = new(StringComparer.OrdinalIgnoreCase)
    {
        ".pdf", ".doc", ".docx", ".txt", ".jpg", ".jpeg", ".png"
    };

    /// <summary>
    /// Allowed content types for image uploads (JPEG, PNG, GIF, WEBP).
    /// </summary>
    public static readonly HashSet<string> AllowedImageTypes = new(StringComparer.OrdinalIgnoreCase)
    {
        "image/jpeg",
        "image/png",
        "image/gif",
        "image/webp"
    };

    /// <summary>
    /// Allowed file extensions for image uploads.
    /// </summary>
    public static readonly HashSet<string> AllowedImageExtensions = new(StringComparer.OrdinalIgnoreCase)
    {
        ".jpg", ".jpeg", ".png", ".gif", ".webp"
    };

    /// <summary>
    /// Human-readable description of allowed document types.
    /// </summary>
    public const string AllowedDocumentTypesDescription = "PDF, DOC, DOCX, TXT, JPG, PNG";

    /// <summary>
    /// Human-readable description of allowed image types.
    /// </summary>
    public const string AllowedImageTypesDescription = "JPEG, PNG, GIF, WEBP";

    /// <summary>
    /// Validates if the given content type is a valid document type.
    /// </summary>
    /// <param name="contentType">The MIME content type to validate.</param>
    /// <returns>True if valid, false otherwise.</returns>
    public static bool IsValidDocumentType(string? contentType) =>
        !string.IsNullOrWhiteSpace(contentType) && AllowedDocumentTypes.Contains(contentType);

    /// <summary>
    /// Validates if the given file extension is a valid document extension.
    /// </summary>
    /// <param name="extension">The file extension to validate (should include the dot, e.g., ".pdf").</param>
    /// <returns>True if valid, false otherwise.</returns>
    public static bool IsValidDocumentExtension(string? extension) =>
        !string.IsNullOrWhiteSpace(extension) && AllowedDocumentExtensions.Contains(extension);

    /// <summary>
    /// Validates if the given content type is a valid image type.
    /// </summary>
    /// <param name="contentType">The MIME content type to validate.</param>
    /// <returns>True if valid, false otherwise.</returns>
    public static bool IsValidImageType(string? contentType) =>
        !string.IsNullOrWhiteSpace(contentType) && AllowedImageTypes.Contains(contentType);

    /// <summary>
    /// Validates if the given file extension is a valid image extension.
    /// </summary>
    /// <param name="extension">The file extension to validate (should include the dot, e.g., ".jpg").</param>
    /// <returns>True if valid, false otherwise.</returns>
    public static bool IsValidImageExtension(string? extension) =>
        !string.IsNullOrWhiteSpace(extension) && AllowedImageExtensions.Contains(extension);

    /// <summary>
    /// Validates a file for document upload by checking both content type and extension.
    /// </summary>
    /// <param name="contentType">The MIME content type.</param>
    /// <param name="fileName">The file name including extension.</param>
    /// <returns>True if either content type or extension is valid, false otherwise.</returns>
    public static bool ValidateDocumentFile(string? contentType, string? fileName)
    {
        // Check content type first
        if (IsValidDocumentType(contentType))
        {
            return true;
        }

        // Fallback to extension validation
        if (!string.IsNullOrWhiteSpace(fileName))
        {
            var extension = Path.GetExtension(fileName);
            return IsValidDocumentExtension(extension);
        }

        return false;
    }

    /// <summary>
    /// Validates a file for image upload by checking both content type and extension.
    /// </summary>
    /// <param name="contentType">The MIME content type.</param>
    /// <param name="fileName">The file name including extension.</param>
    /// <returns>True if either content type or extension is valid, false otherwise.</returns>
    public static bool ValidateImageFile(string? contentType, string? fileName)
    {
        // Check content type first
        if (IsValidImageType(contentType))
        {
            return true;
        }

        // Fallback to extension validation
        if (!string.IsNullOrWhiteSpace(fileName))
        {
            var extension = Path.GetExtension(fileName);
            return IsValidImageExtension(extension);
        }

        return false;
    }
}
